demos/custom-tabs-oauth/src/main/java/com/google/androidbrowserhelper/demos/customtabsoauth/OAuthManager.java

```diff
@@ -0,0 +1,107 @@
+package com.google.androidbrowserhelper.demos.customtabsoauth;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.net.Uri;
+import android.os.Handler;
+import android.os.Looper;
+import android.util.Log;
+
+import androidx.annotation.NonNull;
+import androidx.browser.customtabs.CustomTabsIntent;
+
+import java.io.IOException;
+import java.util.UUID;
+
+import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;
+
+/**
+ * This class helps managing an OAuth flow. It was created wit the goal of demonstrating how to
+ * use Custom Tabs to handle the authorization flow and is not meant as a complete implementation
+ * of the OAuth protocol. We recommend checking out https://github.com/openid/AppAuth-Android for
+ * a comprehensive implementation of the OAuth protocol.
+ */
+public class OAuthManager {
+    private static final String TAG = "OAuthManager";
+
+    private String mClientId;
+    private String mClientSecret;
+    private String mAuthorizationEndpoint;
+    private String mRedirectUri;
+
+    public interface OAuthCallback {
+        void auth(String accessToken, String scope, String tokenType);
+    }
+
+    public OAuthManager(String clientId, String clientSecret, String authorizationEndpoint,
+                        String redirectUri) {
+        this.mClientId = clientId;
+        this.mClientSecret = clientSecret;
+        this.mAuthorizationEndpoint = authorizationEndpoint;
+        this.mRedirectUri = redirectUri;
+    }
+
+    public void authorize(Context context, String scope) {
+        // Generate a random state.
+        String state = UUID.randomUUID().toString();
+
+        // Save the state so we can verify later.
+        SharedPreferences preferences =
+                context.getSharedPreferences("OAUTH_STORAGE", Context.MODE_PRIVATE);
+        preferences.edit()
+                .putString("OAUTH_STATE", state)
+                .apply();
+
+        // Create an authorization URI to the OAuth Endpoint.
+        Uri uri = Uri.parse(mAuthorizationEndpoint)
+                .buildUpon()
+                .appendQueryParameter("response_type", "code")
+                .appendQueryParameter("client_id", mClientId)
+                .appendQueryParameter("redirect_uri", mRedirectUri)
+                .appendQueryParameter("scope", scope)
+                .appendQueryParameter("state", state)
+                .build();
+
+        // Open the Authorization URI in a Custom Tab.
+        CustomTabsIntent customTabsIntent = new CustomTabsIntent.Builder().build();
+        customTabsIntent.intent.setFlags(FLAG_ACTIVITY_NEW_TASK);
```

Comment:
Do CCTs need this?

Reply:
Not setting this is breaking things - When redirecting after auth, instead of opening MainActivity, it seems the Custom Tab just closes and we end up at the home screen.